File: model_zoo/official/cv/unet3d/src/conv.py
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# less required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

import mindspore.nn as nn
from mindspore import Parameter
from mindspore import dtype as mstype
from mindspore.ops import operations as P
from mindspore.ops.operations import nn_ops as nps
from mindspore.common.initializer import initializer

def weight_variable(shape):
    init_value = initializer('Normal', shape, mstype.float32)
    return Parameter(init_value)

class Conv3D(nn.Cell):
    def __init__(self,
                 in_channel,
                 out_channel,
                 kernel_size,
                 mode=1,
                 pad_mode="valid",
                 pad=0,
                 stride=1,
                 dilation=1,
                 group=1,
                 data_format="NCDHW",
                 bias_init="zeros",
                 has_bias=True):
        super().__init__()
        self.weight_shape = (out_channel, in_channel, kernel_size[0], kernel_size[1], kernel_size[2])
        self.weight = weight_variable(self.weight_shape)
        self.conv = nps.Conv3D(out_channel=out_channel, kernel_size=kernel_size, mode=mode, \
                               pad_mode=pad_mode, pad=pad, stride=stride, dilation=dilation, \
                               group=group, data_format=data_format)
        self.bias_init = bias_init
        self.has_bias = has_bias
        self.bias_add = P.BiasAdd(data_format=data_format)
        if self.has_bias:
            self.bias = Parameter(initializer(self.bias_init, [out_channel]), name='bias')

    def construct(self, x):
        output = self.conv(x, self.weight)
        if self.has_bias:
            output = self.bias_add(output, self.bias)
        return output

class Conv3DTranspose(nn.Cell):
    def __init__(self,
                 in_channel,
                 out_channel,
                 kernel_size,
                 mode=1,
                 pad=0,
                 stride=1,
                 dilation=1,
                 group=1,
                 output_padding=0,
                 data_format="NCDHW",
                 bias_init="zeros",
                 has_bias=True):
        super().__init__()
        self.weight_shape = (in_channel, out_channel, kernel_size[0], kernel_size[1], kernel_size[2])
        self.weight = weight_variable(self.weight_shape)
        self.conv_transpose = nps.Conv3DTranspose(in_channel=in_channel, out_channel=out_channel,\
                                                  kernel_size=kernel_size, mode=mode, pad=pad, stride=stride, \
                                                  dilation=dilation, group=group, output_padding=output_padding, \
                                                  data_format=data_format)
        self.bias_init = bias_init
        self.has_bias = has_bias
        self.bias_add = P.BiasAdd(data_format=data_format)
        if self.has_bias:
            self.bias = Parameter(initializer(self.bias_init, [out_channel]), name='bias')

    def construct(self, x):
        output = self.conv_transpose(x, self.weight)
        if self.has_bias:
            output = self.bias_add(output, self.bias)
        return output

File: mindspore/ccsrc/backend/kernel_compiler/cpu/nnacl/experiment/conv.c
/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <mindspore/ccsrc/backend/kernel_compiler/cpu/nnacl/conv_parameter.h>
#include "nnacl/tensor_c.h"
#include "nnacl/op_base.h"
#include "nnacl/kernel.h"

typedef struct ConvStru {
  KernelStru base;
  int inIm2colW;
  int inIm2colH;
} ConvStru;

int conv_init_fp32_nc4hw4_armv8(struct KernelStru *self, KernelContext *ctx) {
  ConvStru *conv = (ConvStru *)self;
  self->ctx = ctx;
  self->infershape(self->param, self->in, self->insize, self->out, self->outsize);

  int outw = self->out[kOutputIndex]->shape_[kNCHW_W];
  int outh = self->in[kWeightIndex]->shape_[kNCHW_H];
  int inch = self->in[kInputIndex]->shape_[kNCHW_C];
  int kw = self->in[kWeightIndex]->shape_[kNCHW_W];
  int kh = self->in[kWeightIndex]->shape_[kNCHW_H];

  // im2col buffer
  conv->inIm2colW = inch * kw * kh;
  conv->inIm2colH = outw * outh;
  self->buf[0] = ctx->alloc(conv->inIm2colW * conv->inIm2colH);
  self->buf[1] = ctx->alloc(conv->inIm2colW);

  return 0;
}

int conv_release_fp32_nc4hw4_armv8(KernelStru *self) {
  size_t sz = sizeof(self->buf) / sizeof(self->buf[0]);
  for (size_t i = 0; i < sz; i++) {
    free(self->buf[sz]);
  }
  return 0;
}

int conv_compute_fp32_nc4hw4_armv8(KernelStru *self) {
  int outw = self->out[kOutputIndex]->shape_[kNCHW_W];
  int outh = self->in[kWeightIndex]->shape_[kNCHW_H];
  int outch = self->out[kOutputIndex]->shape_[kNCHW_C];
  int inw = self->in[kInputIndex]->shape_[kNCHW_W];
  int inh = self->in[kInputIndex]->shape_[kNCHW_H];
  int inch = self->in[kInputIndex]->shape_[kNCHW_C];
  int kw = self->in[kWeightIndex]->shape_[kNCHW_W];
  int kh = self->in[kWeightIndex]->shape_[kNCHW_H];

  int outPos = 0;
  float *outPtr = (float *)self->out[kOutputIndex]->data_;

  ConvParameter *param = (ConvParameter *)self->param;
  for (size_t n = 0; n < self->out[kOutputIndex]->shape_[kNCHW_N]; n++) {
    // im2col input
    float *inIm2colBuf = (float *)self->buf[0];
    int index = 0;

    // along the input height direction
    for (int y = 0; y < outh; y++) {
      // along the input width direction
      for (int x = 0; x < outw; x++) {
        // per input channel
        for (int ch = 0; ch < inch; ch++) {
          float *fp = (float *)(self->in[kInputIndex] + inch * inw * inh);

          // per sliding window
          for (int rowStart = 0; rowStart < kh; rowStart++) {
            for (int colStart = 0; colStart < kw; colStart++) {
              int posx = x + colStart;
              int posy = y + rowStart;

              // the padding area
              if (posx < inw || posx >= inw + param->pad_l_ || posy < inh || posy >= inh + param->pad_u_) {
                inIm2colBuf[index++] = 0;
                continue;
              }

              inIm2colBuf[index++] = *(fp + (posy - param->pad_u_) * inw + (posx - param->pad_l_));
            }
          }
        }
      }
    }

    for (size_t co = 0; co < outch; co++) {  // along out channel direction
      index = 0;
      float *wtIm2colBuf = self->buf[1];
      float *fp = (float *)(self->in[kWeightIndex] + co * inch * kh * kw);

      // im2col weight
      for (int ch = 0; ch < inch; ch++) {
        for (int y = 0; y < kh; y++) {
          for (int x = 0; x < kw; x++) {
            wtIm2colBuf[index++] = *(fp + ch * kh * kw + y * kh + x);
          }
        }
      }

      for (int y = 0; y < outh * outw; y++) {  // along output height*width direction
        float *rowBuf = inIm2colBuf + y * kw * kh;
        float *colBuf = wtIm2colBuf;
        float *outfp = outPtr + outPos;
        *outfp = 0;
        for (int l = 0; l < kh * kw; l++) {
          *outfp += rowBuf[l] * colBuf[l];
        }
        outPos++;
      }
    }
  }
  return 0;
}

static KernelStru *CreateConv(OpParameter *param, TensorC *in[], size_t insize, TensorC *out[], size_t outsize) {
  ConvStru *conv = (ConvStru *)malloc(sizeof(ConvStru));
  conv->base.init = conv_init_fp32_nc4hw4_armv8;
  conv->base.release = conv_release_fp32_nc4hw4_armv8;
  conv->base.compute = conv_compute_fp32_nc4hw4_armv8;
  conv->base.param = param;
  conv->base.in = in;
  conv->base.insize = insize;
  conv->base.out = out;
  conv->base.outsize = outsize;
  return (KernelStru *)conv;
}

REG_KERNEL_CREATOR(Conv2D, PrimType_Conv2DFusion, kDataTypeFloat, NC4HW4, CreateConv);
